### PR TITLE
Task processor framework

### LINKED ISSUE #6742: Turn POC into production code

### PR DESCRIPTION
**Pull request by lpetrovic05**
closes #6742

[proposal doc](https://www.notion.so/swirldslabs/Platform-Component-Rules-Q2-2023-ef93c99cae674b23b4844f200c7b0e9e)

### CODE REVIEW

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/utility/MultiHandler.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import com.swirlds.common.threading.interrupt.InterruptableConsumer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Invokes different handlers based on the type of the object.
+ */
+public class MultiHandler {
```
Comment (cody-littley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/utility/MultiHandler.java:27]: Please add nullity annotations for this class.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/TaskProcessor.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ */
+public interface TaskProcessor {
+
+    /**
+     * Returns a map of references to the task processing methods. The key is the type of the task and the value is a
+     * reference to the method that processes the task.
+     *
+     * @return a map of references to the task processing methods
+     */
+    Map<Class<?>, InterruptableConsumer<?>> getProcessingMethods();
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/TaskProcessor.java:46]:
```suggestion
    @NonNull Map<Class<?>, InterruptableConsumer<?>> getProcessingMethods();
```

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/TaskProcessors.java [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+ *     <li>Create a {@link TaskProcessors} instance, declaring the processors in the constructor</li>
+ *     <li>Request any submitter if other components have dependencies on task processors declared</li>
+ *     <li>Add implementations for processors declared in the constructor</li>
+ *     <li>Start the framework</li>
+ *     <li>Stop the framework when done</li>
+ * </ol>
+ */
+public class TaskProcessors {
+    /** the different parts of task processors */
+    private final Map<Class<? extends TaskProcessor>, ProcessorParts> parts = new HashMap<>();
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/TaskProcessors.java:48]: Can you add a description of the key and value in this map to the javadoc? Understanding this map is super important for understanding the class as a whole.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/TaskProcessors.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+            }
+
+            final QueueThread<Object> queueThread = defaultConfiguration
+                    .copy()
+                    .setThreadName(processorParts.getConfig().name())
+                    .setQueue(processorParts.getQueue())
+                    .setHandler((InterruptableConsumer<Object>) handler)
+                    .build(true);
+            processorParts.setQueueThread(queueThread);
+        }
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/TaskProcessors.java:176]: How does this code behave if you don't provide it with an implementation for a particular task processor type?

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/TaskProcessors.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+ * do so via the submitter created by this framework. The steps to using this framework are:
+ * <ol>
+ *     <li>Create a {@link TaskProcessors} instance, declaring the processors in the constructor</li>
+ *     <li>Request any submitter if other components have dependencies on task processors declared</li>
+ *     <li>Add implementations for processors declared in the constructor</li>
+ *     <li>Start the framework</li>
+ *     <li>Stop the framework when done</li>
+ * </ol>
+ */
+public class TaskProcessors {
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/TaskProcessors.java:46]: This should extend `Lifecycle`. It only needs to support the lifecycle phases `NOT_STARTED`, `STARTED`, and `STOPPED`.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/TaskProcessors.java [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+     * Get the queue thread for the given task processor class.
+     *
+     * @param processorClass
+     * 		the class of the task processor to get the queue thread for
+     * @return the queue thread for the given task processor class
+     */
+    public @NonNull QueueThread<?> getQueueThread(@NonNull final Class<? extends TaskProcessor> processorClass) {
+        throwIfNotStarted();
+        return parts.get(processorClass).getQueueThread();
+    }
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/TaskProcessors.java:190]: Do we need to expose this? Or is this something temporary for backwards compatibility?

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/TaskProcessors.java [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        if (started) {
+            throw new IllegalStateException("Cannot perform this operation after starting");
+        }
+    }
+
+    private void throwIfNotStarted() {
+        if (!started) {
+            throw new IllegalStateException("Cannot perform this operation before starting");
+        }
+    }
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/TaskProcessors.java:210]: Once you extend `Lifecycle` there are some convenient default helper methods such as `throwIfNotInPhase(NOT_STARTED)`.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/internal/QueueSubmitter.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.concurrent.BlockingQueue;
+
+/**
+ * A dynamic proxy that submits tasks to a queue for any method that gets called
+ */
+public class QueueSubmitter implements InvocationHandler {
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/internal/QueueSubmitter.java:27]: nullity annotations for class

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/TaskProcessors.java (hunk 5 above)
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/componentframework/TaskProcessors.java:46]: I wonder if we should consider another name. I know what this does, but the name is not extremely descriptive. Perhaps something like `TaskProcessorBuilder`... we should think on this and discuss.

**Review by hendrikebbers — CHANGES_REQUESTED** [review]
I do not know if that goes in the right direction. I'm afraid that this changes go in a direction that makes platform concurrency more and more complex and kind of incompatible with the java concurrency api. The current draft is not detailed enough for me to understand everything. I would like to know more before approving this PR. I do not want to be a blocker on that but try to help to move the core concurrency in a good direction. Adding @jsync-swirlds since he is interested in that topic, too.

**Review by hendrikebbers — DISMISSED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/utility/MultiHandler.java (hunk 1 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/utility/MultiHandler.java:27]: added

**Comment by jsync-swirlds** [comment]
I agree with Hendrik here.  This entire set of changes makes me uneasy.  Concurrency and multi-thread code is one of the hardest things to get right, and we seem to be trying to create a full _replacement_ for `java.util.concurrent` rather than using the excellent facilities of that API.  The more we can move **away** from DIY concurrency and use what exists in the java platform, the better off we are (and the more benefit we get from the constant evolution of the Java platform).  Most of what we're trying to do here is already in the Java Concurrency APIs, it just looks a little different.
> I do not know if that goes in the right direction. I'm afraid that this changes go in a direction that makes platform concurrency more and more complex and kind of incompatible with the java concurrency api. The current draft is not detailed enough for me to understand everything. I would like to know more before approving this PR. I do not want to be a blocker on that but try to help to move the core concurrency in a good direction. Adding @jsync-swirlds since he is interested in that topic, too.

**Comment by jsync-swirlds** [comment]
Apologies for accidentally closing this.  That was entirely unintentional.
Thanks, @poulok for fixing my error.

**Comment by rbair23** [comment]
I had a chance to look through the PR today. I think I understand what the intent is and what we're trying to accomplish. I agree with @hendrikebbers that I don't think this goes in the right direction, but maybe for a different reason. And I agree with the concept from @jsync-swirlds that we want to use JDK APIs where possible and not needlessly wrap them, though I don't think this PR does that. And knowing @lpetrovic05 I know he would agree too, that we don't want to DIY what we don't have to.

The basic intent of the design is to decouple Component A from Component B. Suppose A is going to do some work, and needs to publish some items to B, so that B can do it's work. We have this pub/sub pattern all throughout our code today. We have a few ways that we can present the coupling between A and B. Suppose A is going to create `McGuffin`s and pass them to B:
  - We could pass B into A's constructor, and have on B a `submitMcGuffin(McGuffin m)` method on B.
  - We could pass a `BlockingQueue<McGuffin>` into the constructor of A that is tied to B.
  - We could pass a Flow/Pipe/Processor/etc. instance into A that is tied to B.
  - We could pass a `Consumer<McGuffin>` into A that is tied to B.
  - We could pass a `McGuffinProcessor` into A that is tied to B.
  - We could pass a `ProcessorRegistry` into A, from which A can extract a `McGuffinProcessor` that is tied to B

Or other variations of the above. In all these cases I'm considering *something* being passed into A's constructor. If A's constructor cannot be used for some reason, then we can have variations of all the above where we set things after construction instead. Although passing into the constructor would be strongly preferred.

The approach taken in the PR is the last, that of using a "Registry". The `TaskProcessor` is the `McGuffinProcessor`, but instead of passing a `McGuffinProcessor` in directly, I guess we would pass in a `TaskProcessors`? I think the `TaskProcessors` is the registry. You create an instance of it, passing in some config that represents the `McGuffinProcessor`. Then you call `addImplementation` separately, and supply a `McGuffinProcessorImpl`, and it looks for the associated config it was given in the constructor and wires it up. Then you call `getSubmitter(McGuffinProcessor.class)` to get back an implementation of `McGuffinProcessor`, which you previously registered as `McGuffinProcessorImpl`. Then you can use that implementation.

Why don't we just take the direct route and pass a `McGuffinProcessor`, or maybe even just a `Consumer<McGuffin>` or, when necessary, an `InterruptableConsumer<McGuffin>`, into the constructor of A? Why do we need the `TaskProcessors` at all? One red flag in the API is that the `TaskProcessor` API has this reflective method on it to get a map of possible types of things that can be consumed. That sits funny. As does the utils class with reflection. That doesn't seem right either.

It seems like this PR is really more about dependency injection than about task processing. Can we just pass the dependency into the constructor instead? Why do we need a registry?

**Comment by lpetrovic05** [comment]
> I agree with Hendrik here. This entire set of changes makes me uneasy. Concurrency and multi-thread code is one of the hardest things to get right, and we seem to be trying to create a full _replacement_ for `java.util.concurrent` rather than using the excellent facilities of that API. The more we can move **away** from DIY concurrency and use what exists in the java platform, the better off we are (and the more benefit we get from the constant evolution of the Java platform). Most of what we're trying to do here is already in the Java Concurrency APIs, it just looks a little different. Note, this shouldn't hold up the PR. It's something we should discuss separately, however.
> 
> > I do not know if that goes in the right direction. I'm afraid that this changes go in a direction that makes platform concurrency more and more complex and kind of incompatible with the java concurrency api. The current draft is not detailed enough for me to understand everything. I would like to know more before approving this PR. I do not want to be a blocker on that but try to help to move the core concurrency in a good direction. Adding @jsync-swirlds since he is interested in that topic, too.

We have a meeting scheduled for this, we can discuss it then.

**Review by artemananiev — APPROVED** [review]
Looks good from platform-data perspective

**Pull request closed without merging**